From cc70ed70b19a712babd834806d6fc700b20c020a Mon Sep 17 00:00:00 2001
From: Emily <hello@emily.moe>
Date: Wed, 22 Jan 2025 23:25:51 +0000
Subject: [PATCH 1/2] =?UTF-8?q?structs:=20Omit=20=E2=80=98free=E2=80=99=20?=
 =?UTF-8?q?field=20from=20=E2=80=98%config-entry=E2=80=99=20on=20libgit2?=
 =?UTF-8?q?=201.9+.?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

* configure.ac: Check for the ‘free’ field of ‘git_config_entry’.
* git/configuration.scm.in (%have-config-entry-free?): New variable.
* git/structs.scm (%config-entry): Omit ‘free’ field conditionally.
---
 configure.ac             | 10 ++++++++++
 git/configuration.scm.in |  4 ++++
 git/structs.scm          |  4 +++-
 3 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index b04ca6b..7f8f0b9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -87,6 +87,16 @@ else
 fi
 AC_SUBST([HAVE_CONFIG_ENTRY_BACKEND_TYPE])
 
+dnl Does 'git_config_entry' have 'free'?  Removed in 1.9.
+AC_CHECK_MEMBER([git_config_entry.free], [], [],
+  [[#include <git2.h>]])
+if test "x$ac_cv_member_git_config_entry_free" = "xyes"; then
+  HAVE_CONFIG_ENTRY_FREE="#true"
+else
+  HAVE_CONFIG_ENTRY_FREE="#false"
+fi
+AC_SUBST([HAVE_CONFIG_ENTRY_FREE])
+
 dnl Does 'git_diff_options' have 'oid_type'? It's new in 1.7.
 AC_CHECK_MEMBER([git_diff_options.oid_type], [], [],
   [[#include <git2.h>]])
diff --git a/git/configuration.scm.in b/git/configuration.scm.in
index e10b195..864b9ff 100644
--- a/git/configuration.scm.in
+++ b/git/configuration.scm.in
@@ -24,6 +24,7 @@
             %have-fetch-options-depth?
             %have-diff-options-oid-type?
             %have-config-entry-backend-type?
+            %have-config-entry-free?
             %have-GIT_OPT_SET_SERVER_CONNECT_TIMEOUT?
             %have-GIT_OPT_SET_HOMEDIR?
             %have-GIT_OPT_SET_USER_AGENT_PRODUCT?))
@@ -48,6 +49,9 @@
 (define %have-config-entry-backend-type?
   @HAVE_CONFIG_ENTRY_BACKEND_TYPE@)
 
+(define %have-config-entry-free?
+  @HAVE_CONFIG_ENTRY_FREE@)
+
 (define %have-GIT_OPT_SET_SERVER_CONNECT_TIMEOUT?
   @HAVE_GIT_OPT_SET_SERVER_CONNECT_TIMEOUT@)
 
diff --git a/git/structs.scm b/git/structs.scm
index beebd11..b35ae6d 100644
--- a/git/structs.scm
+++ b/git/structs.scm
@@ -505,7 +505,9 @@
                      '())
                (include-depth ,unsigned-int)
                (level ,int)                       ;git_config_level_t
-               (free ,(bs:pointer int))
+               ,@(if %have-config-entry-free?
+                     `(free ,(bs:pointer int))
+                     '())
                ,@(if %have-config-entry-backend-type?
                      '()
                      `((payload ,(bs:pointer int))))))) ;removed in 1.8
-- 
2.47.0

